include/vast/Translation/CodeGenDeclVisitor.hpp

```diff
@@ -351,14 +344,38 @@ namespace vast::cg {
                     visit(decl->getBody());
                 }
 
-                // TODO make as pass
-                splice_trailing_scopes(fn);
-
-                auto &last_block = fn.getBlocks().back();
+                auto &fn_blocks  = fn.getBlocks();
+                auto &last_block = fn_blocks.back();
                 auto &ops        = last_block.getOperations();
                 set_insertion_point_to_end(&last_block);
 
-                if (ops.empty() || !is_terminator(ops.back())) {
+                auto last_op = &ops.back();
+
+                // Making sure, that if the operation is enclosed in a trailing
+                // scope, then the termiantor is evaluated in this scope (which
+                // will then be spliced by subsequent pass)
+                auto scope = mlir::dyn_cast< hl::ScopeOp >(ops.back());
+
+                while (scope) {
+                    auto parent = scope->getParentRegion();
+                    if (parent->hasOneBlock()
+                        && parent->back().begin() == --parent->back().end())
+                    {
+                        last_op = get_last_op(scope);
+                        set_insertion_point_to_end(&scope.getBody());
+                        if (last_op)
+                            scope = mlir::dyn_cast< hl::ScopeOp >(last_op);
```

Comment:
`mlir::dyn_cast` returns an object that evaluates to `false` if it fails, so cannot you simply `scope = dyn_cast< ...` and then the while condition will still work?

Also, this can be a local lambda? To avoid re-assigning variables?